// @embabel/runtime-types/index.js
"use strict";
/**
 * @embabel/runtime-types — base types and test helpers for Embabel packs.
 *
 * Pack-specific surfaces (`WorkspaceTools`, namespace shapes) are emitted
 * into each pack's `.embabel/gateway.d.ts` by `embabel-pack sync`. This
 * package supplies only the shared utilities every pack needs.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lens = exports.hours = exports.minutes = exports.seconds = exports.Entity = void 0;
exports.mockGateway = mockGateway;
exports.lensForTest = lensForTest;
exports.entityForTest = entityForTest;
/**
 * Create a typed test stub of a gateway surface.
 *
 * Pack authors call this with their pack's `WorkspaceTools` shape (from
 * the generated `.embabel/gateway.d.ts`) and a partial implementation —
 * usually a small object of `vi.fn()` mocks for the methods the handler
 * under test actually invokes.
 *
 * Returns the input cast to the full type so consumers get autocomplete
 * on every member. Unimplemented members will throw `TypeError` at call
 * time (standard JS behaviour for accessing a missing property as a
 * function) — a clearer signal than silently returning `undefined`.
 *
 * @example
 *   import { mockGateway } from "@embabel/runtime-types";
 *   import type { WorkspaceTools } from "../.embabel/gateway";
 *
 *   const gateway = mockGateway<WorkspaceTools>({
 *     docs: {
 *       documents_get: vi.fn().mockResolvedValue({ revisionId: "r1", body: {...} }),
 *     },
 *   });
 */
function mockGateway(impl) {
    return impl;
}
/**
 * The base class every pack entity type extends.
 *
 * A pack type is a plain class whose async methods are the affordances its
 * in-scope objects expose. The framework hydrates an instance with the entity's
 * fields and injects [gateway] before a method runs, so a method reads the
 * object through `this` and reaches the outside world through `this.gateway`:
 *
 * ```ts
 * export class Movie extends Entity {
 *   imdbId!: string;
 *   title?: string;
 *
 *   // Callable in chat / code-mode as `movie.streaming({ country: "us" })`.
 *   async streaming(args: { country: string }): Promise<unknown> {
 *     return this.gateway.streamingAvailability.getShow({
 *       id: this.imdbId,
 *       country: args.country,
 *     });
 *   }
 * }
 * ```
 *
 * Extending `Entity` is what makes the class a type the host recognises, and it
 * brings [neighbors] for free — graph navigation every entity gets without
 * writing it. There is no `ctx`/`self` plumbing: `this` is the object,
 * `this.gateway` is the context.
 *
 * [gateway] is typed loosely (`GenericGatewayContext`) until `embabel-pack sync`
 * generates the host's `GatewayContext`. For real result types now, declare the
 * slice of the gateway the type uses and read it through a private accessor:
 *
 * ```ts
 * interface MovieGateway {
 *   streamingAvailability: { getShow(a: { id: string; country: string }): Promise<StreamingShow> };
 * }
 * export class Movie extends Entity {
 *   private get api(): MovieGateway { return this.gateway as unknown as MovieGateway; }
 *   async streaming(args: { country: string }): Promise<StreamingShow> {
 *     return this.api.streamingAvailability.getShow({ id: this.imdbId, country: args.country });
 *   }
 * }
 * ```
 */
class Entity {
    /** Stable knowledge-graph id of this node. Set by the host when the object is bound. */
    id;
    /**
     * The gateway, injected by the host before a method runs (and set for you by
     * [entityForTest] in unit tests). Methods reach external APIs and the
     * knowledge graph through it. Protected: it is plumbing, not part of a type's
     * public surface.
     */
    gateway;
    /**
     * Who/what this entity is connected to, related to, or works with in the
     * knowledge graph. Inherited by every pack type, so a new type is graph-
     * navigable for free. Walks outward `hops` (1–3, default 1); optionally filter
     * relationship types (CSV) and cap with `limit`. Returns the connected nodes
     * nearest-first; empty if this node isn't persisted yet.
     */
    neighbors(args = {}) {
        const kg = this.gateway.kg;
        return kg.neighbors({ id: this.id, ...args });
    }
}
exports.Entity = Entity;
/** Seconds helpers so a [CachePolicy] reads in human units but stores plain
 *  seconds — `ttlSeconds: hours(4)`. */
const seconds = (n) => n;
exports.seconds = seconds;
const minutes = (n) => n * 60;
exports.minutes = minutes;
const hours = (n) => n * 3600;
exports.hours = hours;
/**
 * The base class every program lens extends. A lens is a typed, composable
 * retrieval: construct it with its params (`P`), `retrieve()` returns a
 * [LensResult] carrying focus refs and a typed payload (`T`).
 *
 * Retrieval is decoupled from presentation. The author writes only `retrieve()`;
 * the host instantiates the class with the open-time args, injects [gateway],
 * calls `retrieve()`, then bridges the result to focus / present surfaces. The
 * same lens drives entity cards, chat, and generated apps without re-running.
 *
 * Because a lens is a plain class returning a typed value, lenses COMPOSE — one
 * lens constructs and `retrieve()`s another, merging focus and data:
 *
 * ```ts
 * interface BriefView { headline: string; deadlines: Deadline[] }
 *
 * export class BriefingLens extends Lens<{ lookbackDays?: number }, BriefView> {
 *   async retrieve(): Promise<LensResult<BriefView>> {
 *     const rows = await this.gateway.kg.query({ cypher: SELECT, params: "{}" });
 *     return { focus: rows.map(r => ({ id: r.id })), data: { headline, deadlines } };
 *   }
 * }
 *
 * export class DigestLens extends Lens<{}, { brief: BriefView }> {
 *   async retrieve() {
 *     const b = await new BriefingLens({ lookbackDays: 7 }).__bind(this.gateway).retrieve();
 *     return { focus: b.focus, data: { brief: b.data } };
 *   }
 * }
 * ```
 */
class Lens {
    params;
    /** The gateway, injected by the host (via [__bind]) before `retrieve()` runs.
     *  Reaches the knowledge graph and integrations: `this.gateway.kg.query(...)`,
     *  `this.gateway.ai.classify(...)`. Protected: plumbing, not public surface. */
    gateway;
    /**
     * Declarative freshness for this lens. Omit → the host's default (a short
     * coherence window, enough for one opening's views to share a result). Set it
     * when the data ages slowly (`{ ttlSeconds: hours(4) }` for a daily brief) or
     * must stay live (`{ ttlSeconds: 0 }`). A per-result override on
     * [LensResult.cache] wins over this.
     */
    cache;
    constructor(params) {
        this.params = params;
    }
    /** Inject the gateway and return `this`, so construction + binding chain in a
     *  single composing expression: `new ChildLens(p).__bind(this.gateway)`. */
    __bind(gateway) {
        this.gateway = gateway;
        return this;
    }
}
exports.Lens = Lens;
/**
 * Build a program lens for a unit test: a real instance of [ctor] with its
 * params and a [gateway] injected — exactly what the host does at runtime, so
 * `retrieve()` runs unchanged.
 *
 * ```ts
 * const lens = lensForTest(BriefingLens, { lookbackDays: 7 }, mockGateway<Ctx>({
 *   kg: { query: vi.fn().mockResolvedValue([{ id: "n1" }]) },
 * }));
 * const { focus, data } = await lens.retrieve();
 * ```
 */
function lensForTest(ctor, params, gateway) {
    return new ctor(params).__bind(gateway);
}
/**
 * Build a pack entity for a unit test: a real instance of [ctor] with its
 * [fields] set and a [gateway] injected — exactly what the host does at runtime,
 * so the method under test runs unchanged.
 *
 * ```ts
 * const movie = entityForTest(Movie, { imdbId: "tt0114367" }, mockGateway<Ctx>({
 *   streamingAvailability: { getShow: vi.fn().mockResolvedValue({ ... }) },
 * }));
 * await movie.streaming({ country: "us" });
 * ```
 */
function entityForTest(ctor, fields, gateway) {
    const entity = Object.assign(new ctor(), fields);
    entity.gateway = gateway;
    return entity;
}
